app/src/androidTest/java/com/github/sdp/mediato/AuthenticationActivityTest.java

```diff
@@ -0,0 +1,169 @@
+import org.json.JSONObject;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Tests for the authentication activity using the firebase authentication emulator
+ */
+@RunWith(AndroidJUnit4.class)
+public class AuthenticationActivityTest {
+
+    @Rule
+    public ActivityScenarioRule<AuthenticationActivity> testRule = new ActivityScenarioRule<>(AuthenticationActivity.class);
+    private UiDevice device;
+    private AuthenticationActivity activity;
+
+    private FirebaseUser user;
+
+    /**
+     * Logs in the user in the firebase authentication
+     */
+    public void login() {
+
+        // create user json
+        String userJson;
+        String email = "ph@mediato.ch";
+        try {
+            userJson = new JSONObject()
+                    .put("sub", email)
+                    .put("email", email)
+                    .put("email_verified", "true")
+                    .toString();
+        } catch (JSONException e) {
+            throw new IllegalArgumentException(e);
+        }
+
+        // log in user and await result
+        Task<AuthResult> result = FirebaseAuth
+                .getInstance()
+                .signInWithCredential(GoogleAuthProvider
+                        .getCredential(userJson, null));
+        try {
+            Tasks.await(result);
+            user = result.getResult().getUser();
+        } catch (InterruptedException | ExecutionException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Logs out the user in the firebase authentication
+     */
+    public void logout() {
+
+        // log out user and await result
+        Task<Void> result = AuthUI.getInstance()
+                .signOut(ApplicationProvider
+                        .getApplicationContext());
+        try {
+            Tasks.await(result);
+        } catch (InterruptedException | ExecutionException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Starts the emulator and firebase instance, signs-out if user isn't
+     */
+    @Before
+    public void startTests() {
+        init();
+        FirebaseAuth auth = FirebaseAuth.getInstance();
+
+        auth.useEmulator("10.0.2.2", 9099);
+        if (auth.getCurrentUser() != null) {
+            auth.signOut();
+        }
+
+        device = UiDevice.getInstance(getInstrumentation());
+
+        testRule.getScenario().onActivity(activity1 -> activity = activity1);
+    }
+
+    /**
+     * Tests the login one tap button using the emulator
+     *
+     * @throws InterruptedException: for thread.sleep
+     */
+    @Test
+    public void testLogInButtonWorks() throws InterruptedException {
+
+        login();
+        ViewInteraction loginButton = onView(withId(R.id.google_sign_in));
+        loginButton.perform(click());
+
+        Thread.sleep(5000);
```

Comment:
Is there no other way than to wait 5 seconds (like future or callbacks) ? Since it might slow down our test a lot.

Reply:
Unfortunately, we don't have access to the signin intent in those tests, and we can't really access the google one-tap component, so the best way is to use a sleep (ps: reduced it to 3000ms, won't slow down tests very much compared to the already long running time)